### PR TITLE
Support replacing existing media files

### LINKED ISSUE #2922: Option to replace the file of an existing media
n/t

### PR DESCRIPTION
**Pull request by Gravatronics**
Close #2922

I am open for better solutions for https://github.com/WoltLab/WCF/pull/3403/files#diff-b293f4e4dbf3acf0a27e7e999a091c9aR191-R195.

### CODE REVIEW

File: wcfsetup/install/files/lib/page/MediaPage.class.php [hunk 1]
```diff
@@ -92,9 +92,9 @@ public function readData() {
 			'filesize' => $filesize,
 			'showInline' => in_array($mimeType, self::$inlineMimeTypes),
 			'enableRangeSupport' => $this->thumbnail ? true : false,
-			'lastModificationTime' => $this->media->uploadTime,
+			'lastModificationTime' => $this->media->fileUpdateTime ?? $this->media->uploadTime,
 			'expirationDate' => TIME_NOW + 31536000,
-			'maxAge' => 31536000
+			'maxAge' => 31536000,
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/page/MediaPage.class.php:97]: This long of a max age is dangerous if files are able to be replaced.

Likewise the ETag must change after replacement as well.

File: wcfsetup/install/files/js/WoltLabSuite/Core/Media/Manager/Base.js [hunk 2]
```diff
@@ -288,10 +288,28 @@ define(
...
+					p.textContent = media.filename;
+				}
+			}
+			
+			var thumbnail = elByClass('mediaThumbnail', listItem)[0];
+			thumbnail.innerHTML = media.elementTag;
+			// Bust browser cache by adding additional parameter.
+			var imgs = elByTag('img', thumbnail);
+			if (imgs.length) {
+				imgs[0].src += '&refresh=1';
```
Comment (TimWolla) [inline wcfsetup/install/files/js/WoltLabSuite/Core/Media/Manager/Base.js:312]:
```suggestion
				imgs[0].src += '&refresh=' + Date.now();
```

File: wcfsetup/install/files/lib/data/media/MediaAction.class.php [hunk 3]
```diff
@@ -182,6 +188,11 @@ protected function getMediaData($media) {
 				'id' => $media->userID,
 				'title' => $media->username
 			]) : '',
+			// TODO: find better solution
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/data/media/MediaAction.class.php:191]: TODO. This is really ugly.

File: wcfsetup/install/files/lib/data/media/MediaAction.class.php [hunk 4]
```diff
@@ -675,4 +686,87 @@ public function setCategory() {
...
+	}
+	
+	/**
+	 * Replaces the actual file of a media file.
+	 * 
+	 * @return      array
+	 * @since       5.3
+	 */
+	public function replaceFile() {
+		$saveStrategy = new DefaultUploadFileSaveStrategy(self::class, [
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/data/media/MediaAction.class.php:713]:
```suggestion
		$saveStrategy = new DefaultUploadFileSaveStrategy(static::class, [
```

?

File: wcfsetup/install/files/lib/data/media/MediaAction.class.php (hunk 3 above)
Comment (Gravatronics) [reply wcfsetup/install/files/lib/data/media/MediaAction.class.php:191]: That is why I also explicitly added to the message of this pull request that I am open for better solutions :)

May-be we should add a method to UserProfile which relies on an actual template with `{user object=$userProfile}`.

Comment (dtdesign) [reply wcfsetup/install/files/lib/data/media/MediaAction.class.php:191]: I'm not happy about it, but adding bogus extra methods is only moving the issue somewhere else. Moving it (into an actual template) is only masquerading the issue, not solving it.

For the lack of a truly better solution: Keep it.

**Pull request merged**